File: 139. Word Break/139. Word Break.ts
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */
export const wordBreak = function(s: string, wordDict: string[]): boolean {
    const l = s.length;
    const dp: boolean[] = [];
    // dp[i] === true -> s.slice(0, i) in dict
    dp[0] = true;
    for(let i = 1; i <= l; i++){
        dp[i] = false;
        for(let j = 0; j < i; j++){
            if(dp[j]){
                const cut = s.slice(j, i);
                if(wordDict.includes(cut)){
                    dp[i] = true;
                    break;
                }
            }
        }
    }
    return dp[l];
};

File: 139. Word Break/139. Word Break.rs
use std::collections::HashSet;
impl Solution {
    pub fn word_break(s: String, word_dict: Vec<String>) -> bool {
        let words: HashSet<String> = word_dict.into_iter().collect();
        let l = s.len();
        let mut prefix_included: Vec<bool> = vec![false; l + 1];
        prefix_included[0] = true;
        for i in 1..=l {
            for j in 0..i {
                if prefix_included[j] && words.contains(&s[j..i]) {
                    prefix_included[i] = true;
                    break;
                }
            }
        }
        return prefix_included[l];
    }
}
